contracts/axelarnet-gateway/src/state.rs

```diff
@@ -0,0 +1,235 @@
+use axelar_wasm_std::counter::Counter;
+use axelar_wasm_std::{FnExt, IntoContractError};
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Addr, StdError, Storage};
+use cw_storage_plus::{Item, Map};
+use router_api::{ChainName, CrossChainId, Message};
+
+#[cw_serde]
+pub(crate) struct Config {
+    pub chain_name: ChainName,
+    pub router: Addr,
+}
+
+#[cw_serde]
+pub(crate) enum MessageStatus {
+    Approved,
+    Executed,
+}
+
+#[cw_serde]
+pub(crate) struct MessageWithStatus {
+    pub msg: Message,
+    pub status: MessageStatus,
+}
+
+const CONFIG_NAME: &str = "config";
+const CONFIG: Item<Config> = Item::new(CONFIG_NAME);
+
+const COUNTER_NAME: &str = "counter";
+const COUNTER: Counter<u32> = Counter::new(COUNTER_NAME);
+
+const INCOMING_MESSAGES_NAME: &str = "incoming_messages";
+const INCOMING_MESSAGES: Map<CrossChainId, Message> = Map::new(INCOMING_MESSAGES_NAME);
+
+const OUTGOING_MESSAGES_NAME: &str = "outgoing_messages";
+const OUTGOING_MESSAGES: Map<CrossChainId, MessageWithStatus> = Map::new(OUTGOING_MESSAGES_NAME);
+
+#[derive(thiserror::Error, Debug, IntoContractError)]
+pub enum Error {
+    #[error(transparent)]
+    Std(#[from] StdError),
+    #[error("gateway got into an invalid state, its config is missing")]
+    MissingConfig,
```

Comment:
I personally think this sort of thing should be an assert, not an error

Reply:
I usually favour using errors in Rust instead of having to think about whether certain cases can be considered panic-worthy or not (and decide again whenever the related code changes). It's nice to have extra info in case it does get triggered due to some issue. Not a strong opinion for contract code, but we should be consistent everywhere if we decide to switch.

Reply:
you can always use `expect` to output an error message and have the extra info. I prefer to use asserts/expect for things that should never happen. I think this is general best practice, and the purpose of assert/expect/panic. But we are not consistent about this, and there are many places in the code where we return errors for things that should never happen.